### ISSUE TITLE
Get column dtype in expression

### ISSUE THREAD

**Issue by mcrumiller** [pola-rs/polars#4982]
### Problem Description

I've noticed, that sometimes operations can change the dtype of a column. For example, when applying a `pl.duration` to a `pl.Date` object, it converts it to a `pl.Datetime` object:

```python
import polars as pl
from polars import col

pl.DataFrame({'a': [date(2022, 1, 1)]}).select(
    (col('a') + pl.duration(days=1))
)
```

```
shape: (1, 1)
┌─────────────────────┐
│ a                   │
│ ---                 │
│ datetime[ms]        │
╞═════════════════════╡
│ 2022-01-02 00:00:00 │
└─────────────────────┘
```

It would be nice to be able to convert back to the original dtype, without knowing what it was in advance:

```python
pl.DataFrame({'a': [date(2022, 1, 1)]}).select(
    (col('a') + pl.duration(days=1)).cast(col('a').dtype) # convert back to column a's dtype
)
```

**Comment by alexander-beedie** [comment]
As a minor aside (it doesn't invalidate your point ;), I'd suggest only using `pl.duration` for values that include an explicit non-zero nanoseconds component (it creates an `Expr` object). For all other cases you can simply use a vanilla `timedelta`.

```python
pl.DataFrame({'a': [date(2022, 1, 1)]}).select(
    (col('a') + timedelta(days=1))
)
# shape: (1, 1)
# ┌────────────┐
# │ a          │
# │ ---        │
# │ date       │
# ╞════════════╡
# │ 2022-01-02 │
# └────────────┘
```

**Comment by mcrumiller** [comment]
@alexander-beedie thanks for the suggestion. Out of curiosity, does `timedelta` allow you to specify the value found in another column? For example:
 
```python
pl.DataFrame({'a': [date(2022, 1, 1)]}).select(
    (col('a') + timedelta(days=col('b')))
)
```

**Comment by alexander-beedie** [comment]
> @alexander-beedie thanks for the suggestion. Out of curiosity, does `timedelta` allow you to specify the value found in another column? For example:
> 
> ```python
> pl.DataFrame({'a': [date(2022, 1, 1)]}).select(
>     (col('a') + timedelta(days=col('b')))
> )
> ```

No; this is just the regular python `datetime.timedelta`, rather than a cunning polars-aware variant ;)
(So only good for basic offsets - can't read values from an expression/column).

**Comment by mcrumiller** [comment]
> The better solution here would be to preserve the date dtype when adding days (as was recently done for date+timedelta, since https://github.com/pola-rs/polars/pull/4590), rather than making the user do it.

I'm not sure I agree with that suggestion. Is it good practice to have the resulting dtype of an operation depend on the values themselves (not even the dtypes)?

In other words:

`pl.Date + pl.Duration` --> `pl.Date` if the duration granularity <= 1 day
`pl.Date + pl.Duration` --> `pl.Datetime` if the duration granularity > 1 day

This seems a bit wonky to me.

**Comment by ritchie46** [comment]
It gets even worse. The granularity may differ for every element.

In polars, the output of an operation should always depend on the type, never on the data.

**Comment by mcrumiller** [comment]
Hence my desire for the ability to grab the column dtype in an expression :D

**Comment by alexander-beedie** [comment]
> Is it good practice to have the resulting dtype of an operation depend on the values themselves (not even the dtypes)?

I didn't say that the timedelta granularity would _change_ the output dtype, I said that the `Date` dtype would be _preserved_ ;)

This is the same behaviour as adding literal timedeltas (since https://github.com/pola-rs/polars/pull/4590) which is also completely consistent with how Python handles the same thing:

```python
>> date(2022,1,1) + timedelta(days=1)
date(2022, 1, 2)

>> date(2022,1,1) + timedelta(seconds=10)
date(2022, 1, 1)
```
So all I'm suggesting is that `pl.duration` additions are treated the same as timedelta additions.

**Comment by mcrumiller** [comment]
> ```python
> date(2022, 1, 1) + timedelta(days=1)
> date(2022, 1, 2)
>
> date(2022, 1, 1) + timedelta(seconds=10)
> date(2022, 1, 1)
> ```

Now that makes sense. Same as int + float.

**Comment by mcrumiller** [comment]
Encountered another scenario where accessing the column dtype would help, and I can't seem to find a workaround. I implemented a custom numeric -> str format using fprintf-style specifiers like `%5.2f`. If my inputs are series, I can easily check to see if I'm dealing with ints or floats and call e.g. `round()` if it's float, but if the inputs is an expression like `pl.col(...)`, I have no recourse except to auto-cast to float so that `round()` doesn't cause an error, since it's not defined on integer columns.

**Comment by ritchie46** [comment]
This depends on the context the expression runs in. So you would need to pass a schema and a context. I am not sure this is what you wanted?

**Comment by StefanBRas** [comment]
Maybe not speaking for mcrumiller here, but I was thinking a way to specify a dtype as reference to another expression, saying that "The dtype should be the same as the output dtype of this expression". 

Basically exposing the functionality of [GetOutput](https://github.com/pola-rs/polars/blob/a6a21491ee57c7401583d0d74ec19a15f93aef96/polars/polars-lazy/polars-plan/src/dsl/expr.rs#L181) somehow.

**Comment by deanm0000** [comment]
It's interesting that the original example no longer applies here since adding a duration to a date will always return a date now.

For the broader case it probably gets confusing if you're ever using multiple columns like `pl.col("^a\d$")` or something. Otherwise you can use a `pipe` as a work around

```
(
    df
    # Assume several chained with_columns/selects that change initial schema
    .pipe(lambda zz: (
        zz
        .with_columns(
            pl.col(x).expression_that_might_change_dtype().cast(zz.schema[x])
        )
    ))
)
```

**Comment by butterlyn** [comment]
For context, here's the potential use case I had in mind, if hypothetically the dtype was exposed using `polars.Expr.meta.dtype`:

``` python
def process_temporal_column(
    column_expr: pl.Expr,
) -> pl.Expr:
    # if unspecified, make an assumption about the time-of-day
    if column_expr.meta.dtype == pl.Date:
        time_of_day_assumption: pl.Expr = pl.time(hour=8)
        column_expr.dt.combine(time_of_day_assumption)

    # if datetime, leave time-of-day as is

    # some other datetime processing
    ...

    return column_expr

df.with_columns(
    pl.col("date").pipe(process_temporal_column).prefix("processed_"),
    pl.col("datetime").pipe(process_temporal_column).prefix("processed_"),
)
```

**Comment by tpaxman** [comment]
@mcrumiller I had some additional thoughts about potentially adding support for type information when it comes to expressions which I summarized in #18082. Curious to get your feedback as well.

**Comment by deanm0000** [comment]
It couldn't work with a python `if`. It'd have to be a when then expression

```
def process_temporal_column(
    column_expr: pl.Expr,
) -> pl.Expr:
    # if unspecified, make an assumption about the time-of-day
    time_of_day_assumption: pl.Expr = pl.time(hour=8)

    column_expr=(
        pl.when(column_expr.meta.dtype == pl.Date)
        .then(column_expr.dt.combine(time_of_day_assumption))
        .otherwise(column_expr)
        )

    return column_expr
```

**Comment by mcrumiller** [comment]
@deanm0000 couldn't one use a `collect_schema()` and get the dtype to use with python, similar to how we can get the column names?

**Comment by deanm0000** [comment]
@mcrumiller of course but then you don't need the `Expr.dtype` or `Expr.meta.dtype` method at all. To go back to the original example

```
pl.DataFrame({'a': [date(2022, 1, 1)]}).select(
    (col('a') + pl.duration(days=1)).cast(col('a').dtype) # convert back to column a's dtype
)
```

could just be
```
pl.DataFrame({'a': [date(2022, 1, 1)]}).pipe( lambda df: df.select(
    (col('a') + pl.duration(days=1)).cast(df.schema['a']) # convert back to column a's dtype
))
```

**Comment by mcrumiller** [comment]
Ahh indeed. To be fair, I did write the OP almost 2 years ago. I think now that we have `collect_schema` there is a bit less need for the expression, but I can still see how it could be useful to have dtype dependency as a user-facing item in the polars plan.

**Comment by ritchie46** [comment]
Added in #23167
